### ISSUE TITLE
Netcdf code has bugs

### ISSUE THREAD

**Issue by garrettwrong** [garrettwrong/cql3d_f90#23]
As reported by my test harness and confirmed by hand, building with -fPIC (position independent code) causes the output netcdf file to be different in material ways.  

_This is a bug._ It will block library integration. **I would consider it critical.**

```
[gwright@compx-centos7 cql3d_f90]$ nccmp -dmgfs mst_stand_lh00001.MergedCQL3D_direct_lh-dis.nc ci/
case1/              Dockerfile.centos7  
[gwright@compx-centos7 cql3d_f90]$ nccmp -dmgfs mst_stand_lh00001.MergedCQL3D_direct_lh-dis.nc ci/case1/reference.nc 
DIFFER : VARIABLE : nsave : DOES NOT EXIST IN "ci/case1/reference.nc"
DIFFER : LENGTHS : ATTRIBUTE : units : VARIABLE : time : 7 <> 4 : VALUES : "seconds" : "none"
DIFFER : VARIABLE : time : ATTRIBUTE : units : VALUES : "seconds" <> "none"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "long_name" IN FILE "mst_stand_lh00001.MergedCQL3D_direct_lh-dis.nc"
DIFFER : NUMBER OF ATTRIBUTES : VARIABLE : time : 1 <> 2
```

@fmpoli68

**Comment by garrettwrong** [comment]
Confirmed existing in original cql3d_190309 code, by flipping these then doing totally clean builds:

```
#COMPILE   = $(COMPILER) -c $(CSPECIAL) $(DEBUG) -I $(INCLUDE) # or use $(OPTIMIZE)
COMPILE   =  $(COMPILER) -fPIC -c $(CSPECIAL) $(OPTIMIZE) -I $(INCLUDE) # or use $(DEBUG)
```

```
[gwright@compx-centos7 git_cql3d_190309_og]$ nccmp -dfgs *.nc
DIFFER : VARIABLE : nsave : DOES NOT EXIST IN "mst_stand_lh00001.MergedCQL3D_direct_lh-dis.nopic.nc"
DIFFER : LENGTHS : ATTRIBUTE : units : VARIABLE : time : 4 <> 7 : VALUES : "none" : "seconds"
DIFFER : VARIABLE : time : ATTRIBUTE : units : VALUES : "none" <> "seconds"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "long_name" IN FILE "mst_stand_lh00001.MergedCQL3D_direct_lh-dis.pic.nc"
DIFFER : NUMBER OF ATTRIBUTES : VARIABLE : time : 2 <> 1
```

Sorry I did not catch this earlier, but I was mainly on OSX, where disabling PIC is actually harder...
These fields may not be "super important", but the nature of what that difference implies is happening in the code causes me to worry that this is just a symptom of other problems.

**Comment by ypetrov** [comment]
What is the error message?
It looks like 'name' was absent in earlier *.nc file (which is generated 
during run),
but this is normal - we add new variables into subr. netcdfrw2, as the 
code is developed.
I cannot understand - what is the error that stops the run?
--Yuri

**Comment by garrettwrong** [comment]
it doesn't stop the run. Both ways run.. with different resulting output files.

**Comment by garrettwrong** [comment]
Do you expect your code to make those fields.  If you do, there is a bug.  If you do not, there is a bug.

**Comment by ypetrov** [comment]
'time' should always appear in *.nc file. 
'nsave' not always - depends on settings in cqlinput.
Unless these arrays appear to be corrupted in *.nc file, I don't see any bug.

**Comment by garrettwrong** [comment]
the cqlinput was identical.  The code was identical. The only difference is that PIC moves things around in memory, so if you have code that is not using the memory you think it is, you get different behavior.  This code gets difference behavior.

**Comment by ypetrov** [comment]
can you make ncdump of *.nc files from the two runs (with and w/o fPIC)? I want to compare them side by side.

**Comment by ypetrov** [comment]
probably pressed "closed" by mistake.

**Comment by garrettwrong** [comment]
i cleaned them. you should learn to regenerate them. if you do not wish to regenerate them, the automated system has kept a copy here:

https://gitlab.com/geebdubya/cql3d_f90/-/jobs/227398856/artifacts/download

which you can compare to the reference.nc in master.

I included the differences above.

**Comment by ypetrov** [comment]
I think I know the problem. The value of nsavet is not always initialized, so it could be not an integer at all. I am working on it.

**Comment by garrettwrong** [comment]
excellent find, thanks!

**Comment by ypetrov** [comment]
Do I need to "approve" something so that my two updates from yuri_190607 are merged into master?  Did they work for resolving the issue?

**Comment by garrettwrong** [comment]
#24 is merged, you did a nice job with that request, thanks.  I will rebase this branch and push it again, fingers crossed.

**Comment by garrettwrong** [comment]
@ypetrov , this did not resolve.  Did you have any other insights?

I tried to quickly turn on some aggressive debug compile flags, but they do not work with the code out of the box; pretty much because they die on code that I have complained about. Shame, can try more later.

Thanks

**Comment by ypetrov** [comment]
I guess you are talking about -fPIC tests failure.
My only suggestion is to look more into this statement in netcdfrw2:
   istatus= NF_DEF_DIM(ncid, 'time',NF_UNLIMITED,tdim)
I wonder how NF_UNLIMITED is treated in terms of memory,
and how it can be affected by -fPIC flag.

No other ideas at the moment.
Which Debug flags are failing?
I use many run-time flags :
Check array and string bounds,
Check uninitialized variables,
check descriptor data type,
check descriptor data size,
check for actual arguments using tempor. storage.

Note: I cannot use -fPIC under my Intel Compiler in Windows PC.
 From what I read, it is a Linux-only flag (not clear to me - why not 
Windows or Mac?)

Yuri

**Comment by garrettwrong** [comment]
I can take a look, but I think before we imply that netcdf and every library that uses it is broken, we might reflect,  and consider that there isn't any evidence of other people using netcdf blaming their library for something similar, and also more pragmatically, that there is software I am using to compare your netcdf files that must use unlimited vars without any problem.

Windows is not my concern.  In 2019, using windows does not stop developers from doing anything. 
 Virtually any nix-ish system not in a museum has PIC or eqv as a pillar of the operating system.  OSX is PIC by default (they call it something else, but that is just a darwinism, like ld (.so) vs dylib).

I have provided a container that is a completely working linux environment based on Bob's as part of this repository.

**Comment by garrettwrong** [comment]
since yuri is working on cleaning up the nml_dertype branch, I will take this (tomorrow).

**Comment by ypetrov** [comment]
Updated 10 files, see "updates for setup0%".
Am I supposed to request a review for this update?
--Yuri

**Comment by garrettwrong** [comment]
@ypetrov ,No, its already in a request area, I see it.  Looking now.

**Comment by garrettwrong** [comment]
@ypetrov , I turned on implicit none in the netcdfrfw2 file and after a few minutes found the issue. (That is why we want it!).

NCINT was used in place of NCLONG (simple typo im sure), but NCINT didn't exist, so it was implicitly defined and uninitialized.  Changing to NCLONG immediately resolved the problem afaict.

Currently the only diff I get is the following, which corresponded exactly to the change.
[gwright@compx-centos7 cql3d_f90]$ nccmp -dmgfs mst_stand_lh00001.MergedCQL3D_direct_lh-dis.nc PIC
DIFFER : VARIABLE : nsave : TYPE : INT <> DOUBLE
DIFFER : VARIABLE : nsave : POSITION : [0] : VALUES : -2147483647 <> 9.96921e+36

🙈

**Comment by ypetrov** [comment]
Thank you!
Maybe we should add implicit none to every "conversion" subroutine,
they begin at line ~5156 in netcdfrw2 ?

Yuri

Quoting Garrett Wright <notifications@github.com>:

**Comment by garrettwrong** [comment]
The whole code base should be implicit none. Already issue #2 on my list of stuff from last month... but right now it is not a priority.

**Comment by garrettwrong** [comment]
closed by #22
